src/components/Popover/Popover.js

```diff
@@ -1,44 +1,172 @@
+/* eslint-disable react/no-find-dom-node */
 import React from "react";
+import { findDOMNode } from "react-dom";
 import PropTypes from "prop-types";
 import styled from "styled-components";
 import { prop } from "styled-tools";
+import Popper from "popper.js";
 import as from "../../enhancers/as";
 import Hidden from "../Hidden";
-import Perpendicular from "../Perpendicular";
 import Box from "../Box";
+import PopoverArrow from "./PopoverArrow";
 
-const HiddenPerpendicularBox = Hidden.as([Perpendicular, Box]);
+const HiddenBox = Hidden.as(Box);
 
-const Component = props => (
-  <HiddenPerpendicularBox
-    id={props.popoverId}
-    {...props}
-    expand={props.expand === true ? props.pos : props.expand}
-    slide={props.slide === true ? props.pos : props.slide}
-  />
-);
+class Component extends React.Component {
+  state = {
+    placement: this.props.placement,
+    translateX: undefined,
+    translateY: undefined,
+    originX: undefined,
+    originY: undefined
+  };
+
+  getPopover = () => findDOMNode(this);
+
+  modifier = data => {
+    const { placement, offsets, arrowElement, arrowStyles } = data;
+    const { reference, popper } = offsets;
+    const [position] = placement.split("-");
+    const isVertical = ["top", "bottom"].indexOf(position) >= 0;
+
+    const referenceCenter = isVertical
+      ? reference.width / 2
+      : reference.height / 2;
+    const side = isVertical ? "left" : "top";
+    const sideValue = referenceCenter - popper[side];
+
+    if (arrowElement) {
+      const { top, left } = arrowStyles;
+      arrowElement.style.top = isVertical ? "" : `${top}px`;
+      arrowElement.style.left = isVertical ? `${left}px` : "";
+    }
+
+    this.setState({
+      originX: isVertical ? `${sideValue}px - 50%` : 0,
+      originY: !isVertical ? `${sideValue}px - 50%` : 0,
+      translateX: popper.left,
+      translateY: popper.top,
+      placement: data.placement
+    });
+
+    return data;
+  };
+
+  getOptions = () => {
+    const { placement } = this.props;
+    const popover = this.getPopover();
+    const arrowClassName = `.${PopoverArrow.styledComponentId}`;
+    const arrow = popover.querySelector(arrowClassName);
+    return {
+      placement,
+      modifiers: {
+        applyStyle: { enabled: false },
+        arrow: { enabled: !!arrow, element: arrow },
+        flip: { padding: 16 },
+        offset: { offset: `0, ${this.props.gutter}` },
+        setState: {
+          enabled: true,
+          order: 900,
+          fn: this.modifier
+        }
+      }
+    };
+  };
+
+  initPopper = () => {
+    if (!this.popper) {
+      const popover = this.getPopover();
+      this.popper = new Popper(popover.parentNode, popover, this.getOptions());
+    }
+  };
+
+  destroyPopper = () => {
+    if (this.popper) {
+      this.popper.destroy();
+      this.popper = undefined;
```

Comment:
I'm just curious why this is needed.

Reply:
Apparently, we have to destroy popper instance so it won't keep running after the component gets unmounted or another popper instance is initialized (within `componentDidUpdate`), which could cause memory leak.

Reply:
Or are you talking specifically about the last line?

Reply:
Yes, the last line. Won't `destroy` unbind `this.popper`?

Reply:
No. :P